In [2]:
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import pandas as pd

In [3]:
healthcare = pd.read_csv(r'healthcare.csv')
healthcare

ID                Name Age Gender         City Blood Type  \
0     6812       Bobby JacksOn  78      M  Albuquerque         B-   
1     9043        LesLie TErRy  17      F      Albuque         A+   
2     1495         DaNnY sMitH  15   Male      Albuque         A-   
3     4309        andrEw waTtS  32      F      Albuque         O+   
4     8710       adrIENNE bEll  72      M      Albuque        AB+   
...    ...                 ...  ..    ...          ...        ...   
1654  4877        cRaIG cHavEZ  23    NaN    Baltimore        AB-   
1655  6947        Thomas hoode  32    NaN    Baltimore        NaN   
1656  8937    ChristINa barber  19    NaN    Baltimore        AB-   
1657  4870  brANDy FrIeDmaN mD  24    NaN    Baltimore        AB-   
1658  6223          AdAm sMItH  20    NaN    Baltimore         A+   

        Education                Employment Status                 Salary  \
0        Bachelor                         Employed               $80,000    
1       Associate                       Unemployed                    $0    
2             PhD                    Self-employed              $120,000    
3     High School                         Employed               $45,000    
4        Master's                          Retired               $50,000    
...           ...                              ...                    ...   
1654    Associate                       Unemployed                    $0    
1655          PhD                         Employed  $130,000 (Commission)   
1656   Bachelor's  Self-employed (Business Closed)                    $0    
1657          NaN                       Unemployed                    $0    
1658          PhD                         Employed  $110,000 (Commission)   

     Health Condition Credit Score             Date of Admission  
0                Good          720   Wednesday, January 31, 2024  
1           Excellent          NaN      Tuesday, August 20, 2019  
2                Poor          550  Thursday, September 22, 2022  
3           Excellent          800  Wednesday, November 18, 2020  
4             Average          680    Monday, September 19, 2022  
...               ...          ...                           ...  
1654             Poor          NaN                    2020-08-24  
1655             Good          740                    2023-07-17  
1656             Poor          450                    2021-12-28  
1657          Average          400                    2023-12-21  
1658             Good          770                    2020-12-23  

[1659 rows x 12 columns]

### Explaratory data analysis to display basic information

In [4]:
healthcare.head()


ID           Name Age Gender         City Blood Type    Education  \
0  6812  Bobby JacksOn  78      M  Albuquerque         B-     Bachelor   
1  9043   LesLie TErRy  17      F      Albuque         A+    Associate   
2  1495    DaNnY sMitH  15   Male      Albuque         A-          PhD   
3  4309   andrEw waTtS  32      F      Albuque         O+  High School   
4  8710  adrIENNE bEll  72      M      Albuque        AB+     Master's   

  Employment Status     Salary Health Condition Credit Score  \
0          Employed   $80,000              Good          720   
1        Unemployed        $0         Excellent          NaN   
2     Self-employed  $120,000              Poor          550   
3          Employed   $45,000         Excellent          800   
4           Retired   $50,000           Average          680   

              Date of Admission  
0   Wednesday, January 31, 2024  
1      Tuesday, August 20, 2019  
2  Thursday, September 22, 2022  
3  Wednesday, November 18, 2020  
4    Monday, September 19, 2022

In [5]:
healthcare.tail()


ID                Name Age Gender       City Blood Type   Education  \
1654  4877        cRaIG cHavEZ  23    NaN  Baltimore        AB-   Associate   
1655  6947        Thomas hoode  32    NaN  Baltimore        NaN         PhD   
1656  8937    ChristINa barber  19    NaN  Baltimore        AB-  Bachelor's   
1657  4870  brANDy FrIeDmaN mD  24    NaN  Baltimore        AB-         NaN   
1658  6223          AdAm sMItH  20    NaN  Baltimore         A+         PhD   

                    Employment Status                 Salary Health Condition  \
1654                       Unemployed                    $0              Poor   
1655                         Employed  $130,000 (Commission)             Good   
1656  Self-employed (Business Closed)                    $0              Poor   
1657                       Unemployed                    $0           Average   
1658                         Employed  $110,000 (Commission)             Good   

     Credit Score Date of Admission  
1654          NaN        2020-08-24  
1655          740        2023-07-17  
1656          450        2021-12-28  
1657          400        2023-12-21  
1658          770        2020-12-23

In [6]:
healthcare.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1659 entries, 0 to 1658
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   ID                 1659 non-null   int64 
 1   Name               1659 non-null   object
 2   Age                1659 non-null   object
 3   Gender             30 non-null     object
 4   City               1659 non-null   object
 5   Blood Type         1651 non-null   object
 6   Education          1442 non-null   object
 7   Employment Status  1659 non-null   object
 8   Salary             1659 non-null   object
 9   Health Condition   1614 non-null   object
 10  Credit Score       1155 non-null   object
 11  Date of Admission  1659 non-null   object
dtypes: int64(1), object(11)
memory usage: 155.7+ KB


### Identifying duplicates and dropping

In [7]:
healthcare.duplicated().sum()

3

In [8]:
healthcare = healthcare.drop_duplicates()

### Missing values

In [9]:
healthcare.isnull().sum()

ID                      0
Name                    0
Age                     0
Gender               1626
City                    0
Blood Type              8
Education             216
Employment Status       0
Salary                  0
Health Condition       45
Credit Score          504
Date of Admission       0
dtype: int64

### Unique values at the base level to identify anomalies

In [10]:
healthcare['Age'].unique()

array(['78', '17', '15', '32', '72', '91', '52', '56', '43', '20',
       'Unknown', '46', '16', '54', '22', '49', '25', '47', '88', '27',
       '36', '1', '58', '42', '50', '3', '21', '19', '95', '24', '41',
       '51', '18', '8', '64', '30', '2', '39', '45', '31', '0', '23',
       '28', '35', '26', '40', '94', '33', '57', '55', '29', '53', '66',
       '5', '74', '69', '37', '59', '34', '38', '44', '14', '9', '65',
       '90', '48', '62', '60', '4', '70', '10', '11', '63', '6', '73',
       '75', '68', '101', '13', '77', '92', '71', '82', '61', '12', '84',
       '76', '67', '81', '89', '83'], dtype=object)

In [11]:
healthcare['Gender'].unique()

array(['M', 'F', 'Male', 'Female', nan], dtype=object)

In [12]:
healthcare['City'].unique()

array(['Albuquerque', 'Albuque', 'ALBUQUERQUE', 'Atlanta', 'Atl',
       'Baltimore', 'Balti'], dtype=object)

In [13]:
healthcare['Blood Type'].unique()

array(['B-', 'A+', 'A-', 'O+', 'AB+', 'AB-', 'B+', 'O-', nan],
      dtype=object)

In [14]:
healthcare['Education'].unique()

array(['Bachelor', 'Associate', 'PhD', 'High School', "Master's", nan,
       "Bachelor's", 'High School (GED)'], dtype=object)

In [15]:
healthcare['Employment Status'].unique()

array(['Employed', 'Unemployed', 'Self-employed', 'Retired', 'Student',
       'Freelance', 'Student (Part-time)', 'Retired (Early)',
       'Employed (Gig Work)', 'Student (Full-time)',
       'Employed (Part-time)', 'Gig Worker', 'Employed (Contract)',
       'Self-employed (Business Closed)', 'Student (Internship)'],
      dtype=object)

In [16]:
healthcare['Health Condition'].unique()

array(['Good', 'Excellent', 'Poor', 'Average', nan, 'Excellent (?!)'],
      dtype=object)

In [17]:
healthcare['Credit Score'].unique()

array(['720', nan, '550', '800', '680', '400', '750', '600', '790', '537',
       '457', '794', '814', '780', '650', '500', '830', '420', '480',
       '580', '810', '700', '690', '620', '820', '710', '520', '770',
       '740', '840', '450', '730', '760', '670', '640', '850', '610',
       '590', '540', '380 (Typo)', '630', '880', '660'], dtype=object)

In [18]:
healthcare.nunique()

ID                   1513
Name                 1656
Age                    91
Gender                  4
City                    7
Blood Type              8
Education               7
Employment Status      15
Salary                 58
Health Condition        5
Credit Score           42
Date of Admission    1108
dtype: int64

In [19]:
# healthcare.select_dtypes(include = 'number')
# healthcare.select_dtypes(include = 'object')

## Data Overview
Total entries: 1659, columns: 12\
Issues Identified: Duplicates, missing values, typos, wrong data types etc.

In [20]:
# healthcare['Name'] = healthcare['Name'].str.title()
# Correct way to set title case for names. Although the above line also works
healthcare.loc[:, 'Name'] = healthcare['Name'].str.title()

## Cleaning process
### Step 1: Gender Column

In [ ]:
healthcare['Gender'] = healthcare['Gender'].replace({
    'M': 'Male',
    'F': 'Female'
})

healthcare['Gender'].fillna('Unknown', inplace=True)
healthcare['Gender'].unique()

### Step 2: City Column

In [ ]:
healthcare['City'] = healthcare['City'].replace({
    'Albuque': 'Albuquerque',
    'ALBUQUERQUE': 'Albuquerque',
    'Atl': 'Atlanta',
    'Balti': 'Baltimore'
})

healthcare['City'].fillna('Unspecified', inplace=True)
healthcare['City'].unique()

### Step 3: Blood Type

In [ ]:
healthcare['Blood Type'].fillna('Unspecified', inplace=True)
healthcare['Blood Type'].unique()

### Step 4: Education Column

In [ ]:
healthcare['Education'] = healthcare['Education'].replace({
    "Bachelor's": 'Bachelor',
    "Master's": 'Masters',
    'High School (GED)': 'High School',
})
healthcare['Education'].fillna('Unspecified', inplace = True)
healthcare['Education'].unique()

### Step 5: Employment Status Column

In [ ]:
# Manual replacing
# healthcare['Employment Status'] = healthcare['Employment Status'].replace({
#     'Student (Part-time)': 'Student',
#     'Employed (Gig Work)': 'Employed',
#     'Retired (Early)': 'Retired',
#     'Student (Full-time)': 'Student',
#     'Employed (Part-time)': 'Employed',
#     'Employed (Contract)': 'Employed',
#     'Self-employed (Business Closed)': 'Self-employed',
#     'Student (Internship)': 'Student'
# })

# Automatic Replacing, Split the column using space attribute then drop the new column, rename any affected entry!!
healthcare[['Employment Status', 'employment status type']] = healthcare['Employment Status'].str.split(n=1, expand=True)
healthcare['Employment Status'] = healthcare['Employment Status'].replace({'Gig': 'Gig Worker'})
healthcare = healthcare.drop(columns ='employment status type')
healthcare['Employment Status'].unique()

### Step 6: Health Condition Column

In [ ]:
healthcare['Health Condition'] = healthcare['Health Condition'].str.strip("(?!) ")
healthcare['Health Condition'].fillna('Unspecified', inplace=True)
healthcare['Health Condition'].unique()

### Step 8: Credit Score

In [27]:
# I had to replace 'None' with '0' then after finding the 'mean' i rewrote the code to replace 'None' with the the 'Mean'!!
healthcare['Credit Score'] = healthcare['Credit Score'].replace({
    'None': None,
    '380 (Typo)': '380',
    None: '472'
}).astype(int)
healthcare['Credit Score'].unique()

# healthcare['Credit Score'].fillna('0', inplace = True)
# healthcare['Credit Score'] = pd.to_numeric(healthcare['Credit Score'], errors='coerce')
# healthcare['Credit Score'] = healthcare['Credit Score'].astype(int)

array([720, 472, 550, 800, 680, 400, 750, 600, 790, 537, 457, 794, 814,
       780, 650, 500, 830, 420, 480, 580, 810, 700, 690, 620, 820, 710,
       520, 770, 740, 840, 450, 730, 760, 670, 640, 850, 610, 590, 540,
       380, 630, 880, 660])

### Step 8: Age

In [28]:
# Replace 'Unknown' with '0', identify mean as 32 from [describe()] the replace the 32
healthcare['Age'] = healthcare['Age'].replace({'Unknown': '32'}).astype(int)
healthcare['Age'].unique()

array([ 78,  17,  15,  32,  72,  91,  52,  56,  43,  20,  46,  16,  54,
        22,  49,  25,  47,  88,  27,  36,   1,  58,  42,  50,   3,  21,
        19,  95,  24,  41,  51,  18,   8,  64,  30,   2,  39,  45,  31,
         0,  23,  28,  35,  26,  40,  94,  33,  57,  55,  29,  53,  66,
         5,  74,  69,  37,  59,  34,  38,  44,  14,   9,  65,  90,  48,
        62,  60,   4,  70,  10,  11,  63,   6,  73,  75,  68, 101,  13,
        77,  92,  71,  82,  61,  12,  84,  76,  67,  81,  89,  83])

### Step 9: Salary

In [29]:
healthcare[["Salary", "Salary Type"]] = healthcare['Salary'].str.split(n=1, expand=True)
healthcare = healthcare.drop(columns = 'Salary Type')

In [30]:
healthcare['Salary'] = healthcare['Salary'].str.strip("$()")
healthcare['Salary'] = healthcare['Salary'].replace({'Missing': '43664'})
healthcare['Salary'] = pd.to_numeric(healthcare['Salary'].str.replace(',', ''), errors='coerce')
healthcare['Salary'] = healthcare['Salary'].astype(int)
healthcare['Salary'].unique()

array([ 80000,      0, 120000,  45000,  50000,  43664,  65000,  30000,
       100000, 135068,  50975, 102073,  25000,  15000,  35000, 150000,
       110000,  48000,  40000,  60000,  75000,  20000,  55000, 130000,
        28000,  90000, 140000,  32000,  70000,  42000,  18000,  85000,
        38000,  95000,  22000,  12000])

In [31]:
# healthcare['Salary'].value_counts().sort_values()

### Step 7: Date of Admission

In [32]:
# Date class
from datetime import datetime
# Function to clean and format dates: a function named clean_date is defined that takes a single parameter, date_str, which is expected to be a string representing a date
def clean_date(date_str):
    try:
        # Try parsing with the first format
        return datetime.strptime(date_str, '%A, %B %d, %Y').strftime('%Y-%m-%d')
    except ValueError:
        try:
            # Try parsing with the second format "YYYY-MM-DD"
            return datetime.strptime(date_str, '%Y-%m-%d').strftime('%Y-%m-%d')
        except ValueError:
            return None

healthcare['Date of Admission'] = healthcare['Date of Admission'].apply(clean_date)

In [33]:
# Convert column to datetime then extract year, month, year
healthcare['Date of Admission'] = pd.to_datetime(healthcare['Date of Admission'])
healthcare['Year'] = healthcare['Date of Admission'].dt.year
healthcare['Month'] = healthcare['Date of Admission'].dt.month
healthcare['Day'] = healthcare['Date of Admission'].dt.day

In [34]:
# If you wanna change the YYYYMMDD columns to int columns
# healthcare['Day'].fillna('0', inplace = True)
# healthcare['Day']=healthcare['Day'].astype(int)

In [35]:
# Drop useless column
healthcare = healthcare.drop(columns = 'Gender')

In [36]:
# All columns in correct datatypes
healthcare.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1656 entries, 0 to 1655
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   ID                 1656 non-null   int64         
 1   Name               1656 non-null   object        
 2   Age                1656 non-null   int32         
 3   City               1656 non-null   object        
 4   Blood Type         1656 non-null   object        
 5   Education          1656 non-null   object        
 6   Employment Status  1656 non-null   object        
 7   Salary             1656 non-null   int32         
 8   Health Condition   1656 non-null   object        
 9   Credit Score       1656 non-null   int32         
 10  Date of Admission  1641 non-null   datetime64[ns]
 11  Year               1641 non-null   float64       
 12  Month              1641 non-null   float64       
 13  Day                1641 non-null   float64       
dtypes: datetime64

In [37]:
# sns.heatmap(healthcare.describe().corr(), annot = True)
# plt.rcParams['figure.figsize']=(4, 4)
# plt.show()
healthcare.describe()

ID          Age         Salary  Credit Score  \
count  1656.000000  1656.000000    1656.000000   1656.000000   
mean   5535.731280    32.942633   43900.978261    615.891304   
min    1010.000000     0.000000       0.000000    380.000000   
25%    3241.750000    23.000000       0.000000    472.000000   
50%    5545.000000    30.000000   36500.000000    630.000000   
75%    7705.250000    41.000000   80000.000000    742.500000   
max    9999.000000   101.000000  150000.000000    880.000000   
std    2609.672281    14.588154   44706.341728    140.522741   

                   Date of Admission         Year        Month          Day  
count                           1641  1641.000000  1641.000000  1641.000000  
mean   2021-10-12 01:37:24.241316352  2021.285801     6.456429    15.720902  
min              2019-05-09 00:00:00  2019.000000     1.000000     1.000000  
25%              2020-06-20 00:00:00  2020.000000     3.000000     8.000000  
50%              2021-09-25 00:00:00  2021.000000     7.000000    16.000000  
75%              2023-01-15 00:00:00  2023.000000     9.000000    23.000000  
max              2024-05-05 00:00:00  2024.000000    12.000000    31.000000  
std                              NaN     1.513548     3.464729     8.702683

In [38]:
# Removing decimal points from floats
pd.set_option('display.float_format', lambda x: '%.f' % x)
healthcare

ID             Name  Age         City   Blood Type    Education  \
0     6812    Bobby Jackson   78  Albuquerque           B-     Bachelor   
1     9043     Leslie Terry   17  Albuquerque           A+    Associate   
2     1495      Danny Smith   15  Albuquerque           A-          PhD   
3     4309     Andrew Watts   32  Albuquerque           O+  High School   
4     8710    Adrienne Bell   72  Albuquerque          AB+      Masters   
...    ...              ...  ...          ...          ...          ...   
1651  3180  Jonathan Miller   20    Baltimore           B+  High School   
1652  1933        Noah Moon   48    Baltimore           O+     Bachelor   
1653  9661    Jose Anderson   23    Baltimore           O-  Unspecified   
1654  4877     Craig Chavez   23    Baltimore          AB-    Associate   
1655  6947     Thomas Hoode   32    Baltimore  Unspecified          PhD   

     Employment Status  Salary Health Condition  Credit Score  \
0             Employed   80000             Good           720   
1           Unemployed       0        Excellent           472   
2        Self-employed  120000             Poor           550   
3             Employed   45000        Excellent           800   
4              Retired   50000          Average           680   
...                ...     ...              ...           ...   
1651          Employed   60000        Excellent           810   
1652           Student       0             Good           472   
1653     Self-employed  120000          Average           660   
1654        Unemployed       0             Poor           472   
1655          Employed  130000             Good           740   

     Date of Admission  Year  Month  Day  
0           2024-01-31  2024      1   31  
1           2019-08-20  2019      8   20  
2           2022-09-22  2022      9   22  
3           2020-11-18  2020     11   18  
4           2022-09-19  2022      9   19  
...                ...   ...    ...  ...  
1651        2023-07-17  2023      7   17  
1652        2020-10-17  2020     10   17  
1653        2023-11-19  2023     11   19  
1654        2020-08-24  2020      8   24  
1655        2023-07-17  2023      7   17  

[1656 rows x 14 columns]

In [39]:
healthcare.to_csv("cleaned_healthcare_dataset.csv", index=False)